src/riak_core_coverage_fsm.erl

```diff
@@ -332,27 +331,42 @@ terminate(Reason, _StateName, _State) ->
 code_change(_OldVsn, StateName, State, _Extra) ->
     {ok, StateName, State}.
 
-plan_callback(Mod, Exports) ->
-    case exports(plan, Exports) of
-        true ->
+%% This is to avoid expensive module_info calls, which were consuming
+%% 30% of the query-path time for small queries
+%%
+%% Instead we try Mod:plan/2, if undef, we define it.  On success or
+%% match error on atoms, we pass on Mod:plan/2 
+
+plan_callback(Mod) ->
+    try
+        Mod:plan(a, b),
+        fun(CoverageVNodes, ModState) ->
+                Mod:plan(CoverageVNodes, ModState) end
+    catch
+        error:undef ->
+            fun(_, ModState) ->
+                    {ok, ModState} end;
+        _:_ -> %% If Mod:plan(a, b) fails on atoms
             fun(CoverageVNodes, ModState) ->
-                    Mod:plan(CoverageVNodes, ModState) end;
-        _ -> fun(_, ModState) ->
-                     {ok, ModState} end
+                    Mod:plan(CoverageVNodes, ModState) end
     end.
 
-process_results_callback(Mod, Exports) ->
-    case exports_arity(process_results, 3, Exports) of
-        true ->
-            fun(VNode, Results, ModState) ->
-                    Mod:process_results(VNode, Results, ModState) end;
-        false ->
+%% This is to avoid expensive module_info calls, which were consuming
+%% 30% of the query-path time for small queries
+%%
+%% Instead we try Mod:process_results/3, if undef, we define it.  On success or
+%% match error on atoms, we pass on Mod:process_results/3
+
+process_results_callback(Mod) ->
+    try
+        Mod:process_results(a, b, c),
```

Comment:
Same comment as above - introduce a `SuccessFun` and use that in the branches that makes sense.